Rules/UseConsistentCasing.cs

```diff
@@ -0,0 +1,235 @@
+                {
+                    // but handle them separately
+                    UnaryExpressionAst unary = oper as UnaryExpressionAst;
+                    if (unary != null) {
+                        string correctCase = "-" + unary.TokenKind.ToString().ToLowerInvariant();
+                        string actualCase = unary.Extent.Text.Substring(0, correctCase.Length);
+                        if (!actualCase.Equals(correctCase, StringComparison.Ordinal))
+                        {
+                            yield return GetDiagnosticRecord(unary, fileName, correctCase);
+                        }
+                    }
+                    else
+                    {
+                        BinaryExpressionAst binary = oper as BinaryExpressionAst;
+                        string correctCase = binary.ErrorPosition.Text.ToLowerInvariant();
+                        if (!binary.ErrorPosition.Text.Equals(correctCase, StringComparison.Ordinal))
+                        {
+                            yield return GetDiagnosticRecord(binary, fileName, correctCase);
+                        }
+                    }
+                }
+            }
+        }
+
+        private DiagnosticRecord GetDiagnosticRecord(Token token, string fileName, string correction)
+        {
+            var extents = new CorrectionExtent[] {
+                new CorrectionExtent(
+                    token.Extent.StartLineNumber,
+                    token.Extent.EndLineNumber,
+                    token.Extent.StartColumnNumber,
+                    token.Extent.EndColumnNumber,
+                    correction,
+                    token.Extent.File,
+                    GetDescription())};
+
+            return new DiagnosticRecord(
+                string.Format(CultureInfo.CurrentCulture, Strings.UseConsistentCasingKeywordError, token.Text, correction),
+                token.Extent,
+                GetName(),
+                DiagnosticSeverity.Information,
+                fileName,
+                correction, // return the keyword case as the id, so you can turn this off for specific keywords...
+                suggestedCorrections: extents);
+        }
+
+        private DiagnosticRecord GetDiagnosticRecord(UnaryExpressionAst unary, string fileName, string correction)
+        {
+            var extent = new ScriptExtent(
+                new ScriptPosition(
+                    unary.Extent.StartScriptPosition.File,
+                    unary.Extent.StartScriptPosition.LineNumber,
+                    unary.Extent.StartScriptPosition.ColumnNumber,
+                    unary.Extent.StartScriptPosition.Line),
+                new ScriptPosition(
+                    unary.Extent.StartScriptPosition.File,
+                    unary.Extent.StartLineNumber,
+                    unary.Extent.StartColumnNumber + unary.TokenKind.ToString().Length + 1, // add one for the -
+                    unary.Extent.StartScriptPosition.Line));
```

Comment:
The issue with creating new positions/extents (and this is something that should be fixed in PowerShell, but in PSSA we have to work with Windows PS in any case) is that they don't get an offset, which makes them much harder to deal with (and I think PSSA also has a bug where it can't suppress diagnostics with offset-less extents).

But I think the better approach here is to create a diagnostic for the whole operator, rather than everything but the `-`. That way we keep the original extent, and we just write over the whole thing and have to juggle less code in the process.

Reply:
There **is no such extent** in the AST (as far as I can find). There's an extent for the whole expression, and an extent for the "child" but nothing for the operator. So if you write like:

```PowerShell
-join <# this uses newlines #> $lines
```

There's no AST extent for "-join". I can get it from they keywords...

Reply:
> There's no AST extent for "-join"

Ah, yeah, damn. That's annoying. I might have a solution for finding the token...

Reply:
Something like this:

```csharp
int i = 0;
while (tokens[i].Extent.StartOffset <= unary.Extent.StartOffset) { i++; }
while ((tokens[i].TokenFlags & TokenFlags.UnaryOperator) == 0) { i++; }
IScriptExtent extent = tokens[i].Extent;
```